### PR TITLE
nri: add protection against nil dereference

### PR DESCRIPTION
**Pull request by Tal-or**
#### What type of PR is this?

/kind bug

#### What this PR does / why we need it:

If `nil` returned from nri call, set the current `LinuxContainerResources` data.

Integration test: Enabling nri in the server config. 
If the protection is valid, tests should pass.

Additional info in the issue itself 
#### Which issue(s) this PR fixes:

https://github.com/cri-o/cri-o/issues/6642
#### Does this PR introduce a user-facing change?

```release-note
None
```

### CODE REVIEW

File: server/container_update_resources.go [hunk 1]
```diff
@@ -31,6 +31,9 @@ func (s *Server) UpdateContainerResources(ctx context.Context, req *types.Update
 		if err != nil {
 			return nil, err
 		}
+		if updated == nil {
+			updated = req.Linux
+		}
```
Comment (klihub) [inline server/container_update_resources.go:36]: This looks good to me.

File: server/suite_test.go [hunk 2]
```diff
@@ -140,6 +140,7 @@ var beforeEach = func() {
 	Expect(err).To(BeNil())
 	serverConfig, err = config.DefaultConfig()
 	Expect(err).To(BeNil())
+	serverConfig.NRI.Enabled = true
```
Comment (klihub) [inline server/suite_test.go:143]: This is probably not a good idea at the moment. We should bump the NRI dependency to commit `807de1e0cffa` first (and adjust the related bits in the configuration). That eliminates the need for a global system-wide NRI configuration file, which was a bad idea to begin with.

@saschagrunert  @Tal-or I can prepare a PR for that. Then we could try on top of that this `serverConfig.NRI.Enabled = true` and see if we can iron out the any remaining related problems in the tests. WDYT ?

Comment (haircommander) [reply server/suite_test.go:143]: agreed. I would say probably having dedicated tests with NRI enabled would be good

Comment (Tal-or) [reply server/suite_test.go:143]: Yes, the test failure is due to the missing configuration file indeed.
@haircommander I can address that, i'll edit the PR accordingly while we're waiting for the dependency bump.

Comment (klihub) [reply server/suite_test.go]: @haircommander, @Tal-or #6648 is the PR which bumps the NRI deps, eliminating the global config file (and it also adjusts the related configuration, documentation, etc. accordingly).

I'm see quite a bit of test failures there, but I see the same ones (BATS tests  42, 45, 46, 47) fail in many of the other recent PRs. Are these known to be broken now ?

Comment (klihub) [reply server/suite_test.go]: @Tal-or @hircommander I tested #6648 with 96cde59d99d8ce0974e6c6fbbefc60d176901471 cherry-picked to see how it goes. 3 of the unit test crashed with `internal/oci/Container` trying to dereference an unset nil `OCI Spec`, all others passed. [Added a fix for that crash](https://github.com/klihub/cri-o/tree/release-test) and am testing now to see if anything else comes up to fix. If things look fine, I'll update #6648 with that fix included.

Comment (Tal-or) [reply server/suite_test.go]: Thank you @klihub are the crashes in `internal/oci/Container` related to the one you mentioned here?
https://github.com/cri-o/cri-o/pull/6645#discussion_r1106421489

Comment (klihub) [reply server/suite_test.go]: @Tal-or Well, they are caused by an unset/`nil` OCI Spec being unconditionally dereferenced by `Spec()`. I've added [a fix](https://github.com/cri-o/cri-o/pull/6648/commits/bd7fb2416547dea477aa75de52e1f76082b8a0b3) for it in #6648.

@haircommander @Tal-or I've tested this rebased on #6648 and unit tests do pass now.

**Comment by haircommander** [comment]
/approve
/override ci/prow/e2e-aws-ovn
/override ci/prow/e2e-gcp-ovn

LGTM, PTAL @cri-o/cri-o-maintainers

**Review by jrvaldes — COMMENTED** [review]
/lgtm

**Comment by jrvaldes** [comment]
@Tal-or  Does this change needs to be backported?

cc @haircommander

**Comment by haircommander** [comment]
I could see it being useful for 1.26, though we'd proably have to backport @klihub 's vendor bump PR first, no?

File: server/container_update_resources_test.go [hunk 3]
```diff
@@ -120,4 +118,36 @@ var _ = t.Describe("UpdateContainerResources", func() {
 			Expect(err).NotTo(BeNil())
 		})
 	})
+
+	t.Describe("UpdateContainerResources with nri enabled", func() {
+		// Prepare the sut
+		BeforeEach(func() {
+			beforeEach()
+			serverConfig.NRI.Enabled = true
```
Comment (klihub) [inline server/container_update_resources_test.go:126]: @Tal-or Wasn't your initial take on this to unconditionally run all unit tests with NRI enabled ? I think that could still be a good idea...

**Comment by klihub** [comment]
> I could see it being useful for 1.26, though we'd proably have to backport @klihub 's vendor bump PR first, no?

@haircommander Yes, we'd need that PR, in particular ab3b401b0f8da3bed9606a66e9103ab6c169d8ad to avoid a nil dereference crash in internal/oci/container.go.Spec().

And I think it could be a good idea to force more/all of the unit tests running with NRI enabled... just like @Tal-or was doing in the original PR.

File: server/container_update_resources_test.go (hunk 3 above)
Comment (Tal-or) [reply server/container_update_resources_test.go:126]: I actually followed an earlier request from @haircommander.
But it's doable indeed

Comment (klihub) [reply server/container_update_resources_test.go:126]: Okay, sorry I missed that. Let's go with what  @haircommander think is the best course of action.

**Comment by Tal-or** [comment]
the ci `integration / test` falling with:
```
not ok 94 ctr oom
# (in test file ./ctr.bats, line 755)
#   `[[ "$output" == *"OOMKilled"* ]]' failed
# time="2023-02-22 17:46:58.343325486Z" level=info msg="Starting CRI-O, version: 1.26.0, git: c7d25501490fe71c82b3c45e73e4a636575556c2(clean)"
# level=info msg="Using default capabilities: CAP_CHOWN, CAP_DAC_OVERRIDE, CAP_FSETID, CAP_FOWNER, CAP_SETGID, CAP_SETUID, CAP_SETPCAP, CAP_NET_BIND_SERVICE, CAP_KILL"
# time="2023-02-22 17:46:58.463978974Z" level=info msg="Starting CRI-O, version: 1.26.0, git: c7d25501490fe71c82b3c45e73e4a636575556c2(clean)"
# level=info msg="Using default capabilities: CAP_CHOWN, CAP_DAC_OVERRIDE, CAP_FSETID, CAP_FOWNER, CAP_SETGID, CAP_SETUID, CAP_SETPCAP, CAP_NET_BIND_SERVICE, CAP_KILL"
# E0222 17:46:58.566785   77995 remote_runtime.go:616] "Status from runtime service failed" err="rpc error: code = Unavailable desc = connection error: desc = \"transport: Error while dialing dial unix /tmp/tmp.PLst70sWWQ/crio.sock: connect: no such file or directory\""
# time="2023-02-22T17:46:58Z" level=fatal msg="getting status of runtime: rpc error: code = Unavailable desc = connection error: desc = \"transport: Error while dialing dial unix /tmp/tmp.PLst70sWWQ/crio.sock: connect: no such file or directory\""
```
But I don't think it's related to this PR's changes

**Comment by klihub** [comment]
> the ci `integration / test` falling with:
> 
> ```
> not ok 94 ctr oom
> # (in test file ./ctr.bats, line 755)
> #   `[[ "$output" == *"OOMKilled"* ]]' failed
> # time="2023-02-22 17:46:58.343325486Z" level=info msg="Starting CRI-O, version: 1.26.0, git: c7d25501490fe71c82b3c45e73e4a636575556c2(clean)"
> # level=info msg="Using default capabilities: CAP_CHOWN, CAP_DAC_OVERRIDE, CAP_FSETID, CAP_FOWNER, CAP_SETGID, CAP_SETUID, CAP_SETPCAP, CAP_NET_BIND_SERVICE, CAP_KILL"
> # time="2023-02-22 17:46:58.463978974Z" level=info msg="Starting CRI-O, version: 1.26.0, git: c7d25501490fe71c82b3c45e73e4a636575556c2(clean)"
> # level=info msg="Using default capabilities: CAP_CHOWN, CAP_DAC_OVERRIDE, CAP_FSETID, CAP_FOWNER, CAP_SETGID, CAP_SETUID, CAP_SETPCAP, CAP_NET_BIND_SERVICE, CAP_KILL"
> # E0222 17:46:58.566785   77995 remote_runtime.go:616] "Status from runtime service failed" err="rpc error: code = Unavailable desc = connection error: desc = \"transport: Error while dialing dial unix /tmp/tmp.PLst70sWWQ/crio.sock: connect: no such file or directory\""
> # time="2023-02-22T17:46:58Z" level=fatal msg="getting status of runtime: rpc error: code = Unavailable desc = connection error: desc = \"transport: Error while dialing dial unix /tmp/tmp.PLst70sWWQ/crio.sock: connect: no such file or directory\""
> ```
> 
> But I don't think it's related to this PR's changes

I see this (and a few  others) happening fairly regularly for other PRs, too.

**Comment by Tal-or** [comment]
The PR is blocker because:
`
Required status check "ci/kata-jenkins" was not set by the expected GitHub app.
`
@haircommander @saschagrunert could you please advise how  it can be resolved?

**Comment by Tal-or** [comment]
unclear to me why `ci/prow/e2e-aws-ovn` which got overridden is still required

**Comment by saschagrunert** [comment]
@Tal-or it's because prow will re-test everything again if another PR merges in between yours and `main`. Let's get this merged now in priority.

**Comment by Tal-or** [comment]
> @Tal-or it's because prow will re-test everything again if another PR merges in between yours and `main`. Let's get this merged now in priority.

I see, thank you for the explanation

**Comment by saschagrunert** [comment]
It should merge now.

**Pull request merged**